File: src/sim/weather/WeatherSystem.ts
import type { Rng } from "../../core/math/Rng";
import type { TerrainGrid } from "../terrain/TerrainGrid";

/**
 * Météo cellulaire (docs/GDD.md §3.4, TDD §4.x) sur une grille grossière
 * (1 cellule = WEATHER_CELL tuiles), cadencée tous les WEATHER_INTERVAL ticks.
 *
 * Boucle de l'eau, entièrement déterministe (stream RNG "weather") :
 *   évaporation au-dessus de l'eau → nuages advectés par un vent qui tourne
 *   lentement → précipitations quand un nuage sature au-dessus des terres →
 *   l'humidité du sol monte (les biomes verdissent) → sans pluie, le sol
 *   revient vers son humidité d'équilibre (baseline).
 *
 * Le pouvoir divin « Pluie » (`RainPower`) ne fait que saturer des nuages :
 * la pluie qui suit est le comportement normal du système.
 */
export const WEATHER_CELL = 8;
export const WEATHER_INTERVAL = 5;

/** Cloudiness above which it rains (or snows on cold ground). */
export const PRECIPITATION_THRESHOLD = 0.62;
/** Soil moisture gained per rained-on tile per weather update. */
const RAIN_MOISTURE = 0.01;
/** Cloud mass lost per rainy update. */
const RAIN_DEPLETION = 0.045;
/** Cloud gained per update over open water. */
const EVAPORATION = 0.035;
/** Fraction of the gap to baseline moisture closed per update (drying). */
const DRY_RATE = 0.002;
/** Passive cloud dissipation per update. */
const DISSIPATION = 0.996;
/** Ground base temperature below which precipitation falls as snow. */
export const SNOW_TEMPERATURE = 0.3;

export class WeatherSystem {
  readonly cellsX: number;
  readonly cellsY: number;
  /** Couverture nuageuse par cellule, [0, 1]. */
  readonly cloud: Float32Array;
  windAngle: number;
  private advectionX = 0;
  private advectionY = 0;
  private readonly rng: Rng;

  constructor(
    private readonly terrain: TerrainGrid,
    baseRng: Rng,
  ) {
    this.cellsX = Math.ceil(terrain.width / WEATHER_CELL);
    this.cellsY = Math.ceil(terrain.height / WEATHER_CELL);
    this.cloud = new Float32Array(this.cellsX * this.cellsY);
    this.rng = baseRng.fork("weather");
    this.windAngle = this.rng.float() * Math.PI * 2;
    for (let i = 0; i < this.cloud.length; i++) {
      this.cloud[i] = this.rng.float() * 0.4;
    }
  }

  cellIndex(cx: number, cy: number): number {
    return cy * this.cellsX + cx;
  }

  cloudAt(cx: number, cy: number): number {
    return this.cloud[this.cellIndex(cx, cy)]!;
  }

  /** Pluie en cours dans cette cellule ? (rendu + gameplay) */
  isRaining(cx: number, cy: number): boolean {
    return this.cloudAt(cx, cy) > PRECIPITATION_THRESHOLD && !this.isCellOverWater(cx, cy);
  }

  /** Neige plutôt que pluie : sol froid (latitude/saison via baseTemperature). */
  isSnowing(cx: number, cy: number): boolean {
    if (!this.isRaining(cx, cy)) return false;
    const { x, y } = this.cellCentreTile(cx, cy);
    return this.terrain.baseTemperature[this.terrain.index(x, y)]! + this.terrain.seasonalOffset < SNOW_TEMPERATURE;
  }

  /** Sature les nuages autour d'un point (utilisé par le pouvoir Pluie). */
  seedClouds(tileX: number, tileY: number, tileRadius: number): void {
    const cx0 = Math.floor((tileX - tileRadius) / WEATHER_CELL);
    const cy0 = Math.floor((tileY - tileRadius) / WEATHER_CELL);
    const cx1 = Math.floor((tileX + tileRadius) / WEATHER_CELL);
    const cy1 = Math.floor((tileY + tileRadius) / WEATHER_CELL);
    for (let cy = Math.max(0, cy0); cy <= Math.min(this.cellsY - 1, cy1); cy++) {
      for (let cx = Math.max(0, cx0); cx <= Math.min(this.cellsX - 1, cx1); cx++) {
        this.cloud[this.cellIndex(cx, cy)] = 1;
      }
    }
  }

  /** Un pas de météo (appelé tous les WEATHER_INTERVAL ticks). */
  update(): void {
    this.driftWind();
    this.advect();

    for (let cy = 0; cy < this.cellsY; cy++) {
      for (let cx = 0; cx < this.cellsX; cx++) {
        const i = this.cellIndex(cx, cy);
        const overWater = this.isCellOverWater(cx, cy);

        if (overWater) {
          this.cloud[i] = Math.min(1, this.cloud[i]! + EVAPORATION);
        } else if (this.cloud[i]! > PRECIPITATION_THRESHOLD) {
          this.cloud[i] = this.cloud[i]! - RAIN_DEPLETION;
          this.rainOnCell(cx, cy);
        }

        this.cloud[i] = this.cloud[i]! * DISSIPATION;
      }
    }

    this.drySoil();
  }

  /** Le vent tourne lentement et aléatoirement (déterministe). */
  private driftWind(): void {
    this.windAngle += (this.rng.float() - 0.5) * 0.15;
  }

  /** Advection : décale la grille de nuages d'une cellule quand le vent a assez soufflé. */
  private advect(): void {
    const speed = 0.35; // cellules par update
    this.advectionX += Math.cos(this.windAngle) * speed;
    this.advectionY += Math.sin(this.windAngle) * speed;
    while (Math.abs(this.advectionX) >= 1) {
      this.shift(Math.sign(this.advectionX), 0);
      this.advectionX -= Math.sign(this.advectionX);
    }
    while (Math.abs(this.advectionY) >= 1) {
      this.shift(0, Math.sign(this.advectionY));
      this.advectionY -= Math.sign(this.advectionY);
    }
  }

  /** Décalage torique de la grille de nuages. */
  private shift(dx: number, dy: number): void {
    const { cellsX, cellsY } = this;
    const source = this.cloud.slice();
    for (let cy = 0; cy < cellsY; cy++) {
      for (let cx = 0; cx < cellsX; cx++) {
        const sx = (cx - dx + cellsX) % cellsX;
        const sy = (cy - dy + cellsY) % cellsY;
        this.cloud[this.cellIndex(cx, cy)] = source[this.cellIndex(sx, sy)]!;
      }
    }
  }

  private rainOnCell(cx: number, cy: number): void {
    const x0 = cx * WEATHER_CELL;
    const y0 = cy * WEATHER_CELL;
    const x1 = Math.min(x0 + WEATHER_CELL, this.terrain.width);
    const y1 = Math.min(y0 + WEATHER_CELL, this.terrain.height);
    for (let y = y0; y < y1; y++) {
      for (let x = x0; x < x1; x++) {
        if (this.terrain.isWater(x, y)) continue;
        const i = this.terrain.index(x, y);
        this.terrain.setMoisture(x, y, this.terrain.moisture[i]! + RAIN_MOISTURE);
      }
    }
  }

  /** Sans pluie, l'humidité revient lentement vers l'équilibre généré. */
  private drySoil(): void {
    const terrain = this.terrain;
    for (let i = 0; i < terrain.moisture.length; i++) {
      const gap = terrain.baselineMoisture[i]! - terrain.moisture[i]!;
      if (Math.abs(gap) < 0.0005) continue;
      const x = i % terrain.width;
      const y = Math.floor(i / terrain.width);
      terrain.setMoisture(x, y, terrain.moisture[i]! + gap * DRY_RATE);
    }
  }

  private isCellOverWater(cx: number, cy: number): boolean {
    const { x, y } = this.cellCentreTile(cx, cy);
    return this.terrain.isWater(x, y);
  }

  private cellCentreTile(cx: number, cy: number): { x: number; y: number } {
    return {
      x: Math.min(this.terrain.width - 1, cx * WEATHER_CELL + WEATHER_CELL / 2),
      y: Math.min(this.terrain.height - 1, cy * WEATHER_CELL + WEATHER_CELL / 2),
    };
  }

  /** Snapshot pour la sauvegarde (v2). */
  serialize(): { cloud: number[]; windAngle: number; advectionX: number; advectionY: number; rngState: number } {
    return {
      cloud: Array.from(this.cloud),
      windAngle: this.windAngle,
      advectionX: this.advectionX,
      advectionY: this.advectionY,
      rngState: this.rng.getState(),
    };
  }

  /** Restauration depuis une sauvegarde (v2). */
  restore(data: { cloud: number[]; windAngle: number; advectionX: number; advectionY: number; rngState: number }): void {
    if (data.cloud.length !== this.cloud.length) {
      throw new Error("Corrupted save: weather grid size mismatch");
    }
    this.cloud.set(data.cloud);
    this.windAngle = data.windAngle;
    this.advectionX = data.advectionX;
    this.advectionY = data.advectionY;
    this.rng.setState(data.rngState);
  }
}

File: unity/ImG.Core/Weather/WeatherSystem.cs
using System;
using ImG.Core.Math;
using ImG.Core.Terrain;

namespace ImG.Core.Weather
{
    /// <summary>
    /// Météo cellulaire (docs/GDD.md §3.4) sur une grille grossière
    /// (1 cellule = <see cref="WeatherCell"/> tuiles), cadencée tous les
    /// <see cref="WeatherInterval"/> ticks. Boucle de l'eau entièrement
    /// déterministe (flux RNG « weather ») : évaporation au-dessus de l'eau →
    /// nuages advectés par le vent → précipitations sur les terres saturées →
    /// l'humidité du sol monte → sans pluie, elle revient vers l'équilibre.
    ///
    /// Portage de <c>src/sim/weather/WeatherSystem.ts</c>. Le tableau de nuages
    /// est en <see cref="float"/> pour reproduire à l'identique la précision des
    /// <c>Float32Array</c> du jeu TypeScript (chaque écriture est arrondie en
    /// float32).
    /// </summary>
    public sealed class WeatherSystem
    {
        public const int WeatherCell = 8;
        public const int WeatherInterval = 5;

        /// <summary>Couverture au-dessus de laquelle il pleut (ou neige sur sol froid).</summary>
        public const double PrecipitationThreshold = 0.62;
        private const double RainMoisture = 0.01;
        private const double RainDepletion = 0.045;
        private const double Evaporation = 0.035;
        private const double DryRate = 0.002;
        private const double Dissipation = 0.996;
        /// <summary>Température de base du sol sous laquelle la pluie tombe en neige.</summary>
        public const double SnowTemperature = 0.3;

        public readonly int CellsX;
        public readonly int CellsY;
        /// <summary>Couverture nuageuse par cellule, [0, 1].</summary>
        public readonly float[] Cloud;
        public double WindAngle;

        private double _advectionX;
        private double _advectionY;
        private readonly Rng _rng;
        private readonly TerrainGrid _terrain;

        public WeatherSystem(TerrainGrid terrain, Rng baseRng)
        {
            _terrain = terrain;
            CellsX = (int)System.Math.Ceiling((double)terrain.Width / WeatherCell);
            CellsY = (int)System.Math.Ceiling((double)terrain.Height / WeatherCell);
            Cloud = new float[CellsX * CellsY];
            _rng = baseRng.Fork("weather");
            WindAngle = _rng.Float() * System.Math.PI * 2;
            for (int i = 0; i < Cloud.Length; i++)
                Cloud[i] = (float)(_rng.Float() * 0.4);
        }

        public int CellIndex(int cx, int cy) => cy * CellsX + cx;

        public double CloudAt(int cx, int cy) => Cloud[CellIndex(cx, cy)];

        /// <summary>Pluie en cours dans cette cellule ? (rendu + gameplay)</summary>
        public bool IsRaining(int cx, int cy) =>
            CloudAt(cx, cy) > PrecipitationThreshold && !IsCellOverWater(cx, cy);

        /// <summary>Neige plutôt que pluie : sol froid (latitude/saison).</summary>
        public bool IsSnowing(int cx, int cy)
        {
            if (!IsRaining(cx, cy)) return false;
            var (x, y) = CellCentreTile(cx, cy);
            return _terrain.BaseTemperature[_terrain.Index(x, y)] + _terrain.SeasonalOffset < SnowTemperature;
        }

        /// <summary>Sature les nuages autour d'un point (utilisé par le pouvoir Pluie).</summary>
        public void SeedClouds(int tileX, int tileY, int tileRadius)
        {
            int cx0 = (int)System.Math.Floor((double)(tileX - tileRadius) / WeatherCell);
            int cy0 = (int)System.Math.Floor((double)(tileY - tileRadius) / WeatherCell);
            int cx1 = (int)System.Math.Floor((double)(tileX + tileRadius) / WeatherCell);
            int cy1 = (int)System.Math.Floor((double)(tileY + tileRadius) / WeatherCell);
            for (int cy = System.Math.Max(0, cy0); cy <= System.Math.Min(CellsY - 1, cy1); cy++)
                for (int cx = System.Math.Max(0, cx0); cx <= System.Math.Min(CellsX - 1, cx1); cx++)
                    Cloud[CellIndex(cx, cy)] = 1f;
        }

        /// <summary>Un pas de météo (appelé tous les <see cref="WeatherInterval"/> ticks).</summary>
        public void Update()
        {
            DriftWind();
            Advect();

            for (int cy = 0; cy < CellsY; cy++)
            {
                for (int cx = 0; cx < CellsX; cx++)
                {
                    int i = CellIndex(cx, cy);
                    bool overWater = IsCellOverWater(cx, cy);

                    if (overWater)
                    {
                        Cloud[i] = (float)System.Math.Min(1, Cloud[i] + Evaporation);
                    }
                    else if (Cloud[i] > PrecipitationThreshold)
                    {
                        Cloud[i] = (float)(Cloud[i] - RainDepletion);
                        RainOnCell(cx, cy);
                    }

                    Cloud[i] = (float)(Cloud[i] * Dissipation);
                }
            }

            DrySoil();
        }

        /// <summary>Le vent tourne lentement et aléatoirement (déterministe).</summary>
        private void DriftWind() => WindAngle += (_rng.Float() - 0.5) * 0.15;

        /// <summary>Advection : décale la grille d'une cellule quand le vent a assez soufflé.</summary>
        private void Advect()
        {
            const double speed = 0.35; // cellules par update
            _advectionX += System.Math.Cos(WindAngle) * speed;
            _advectionY += System.Math.Sin(WindAngle) * speed;
            while (System.Math.Abs(_advectionX) >= 1)
            {
                Shift(System.Math.Sign(_advectionX), 0);
                _advectionX -= System.Math.Sign(_advectionX);
            }
            while (System.Math.Abs(_advectionY) >= 1)
            {
                Shift(0, System.Math.Sign(_advectionY));
                _advectionY -= System.Math.Sign(_advectionY);
            }
        }

        /// <summary>Décalage torique de la grille de nuages.</summary>
        private void Shift(int dx, int dy)
        {
            var source = (float[])Cloud.Clone();
            for (int cy = 0; cy < CellsY; cy++)
            {
                for (int cx = 0; cx < CellsX; cx++)
                {
                    int sx = (cx - dx + CellsX) % CellsX;
                    int sy = (cy - dy + CellsY) % CellsY;
                    Cloud[CellIndex(cx, cy)] = source[CellIndex(sx, sy)];
                }
            }
        }

        private void RainOnCell(int cx, int cy)
        {
            int x0 = cx * WeatherCell;
            int y0 = cy * WeatherCell;
            int x1 = System.Math.Min(x0 + WeatherCell, _terrain.Width);
            int y1 = System.Math.Min(y0 + WeatherCell, _terrain.Height);
            for (int y = y0; y < y1; y++)
            {
                for (int x = x0; x < x1; x++)
                {
                    if (_terrain.IsWater(x, y)) continue;
                    int i = _terrain.Index(x, y);
                    _terrain.SetMoisture(x, y, _terrain.Moisture[i] + RainMoisture);
                }
            }
        }

        /// <summary>Sans pluie, l'humidité revient lentement vers l'équilibre généré.</summary>
        private void DrySoil()
        {
            for (int i = 0; i < _terrain.Moisture.Length; i++)
            {
                double gap = _terrain.BaselineMoisture[i] - _terrain.Moisture[i];
                if (System.Math.Abs(gap) < 0.0005) continue;
                int x = i % _terrain.Width;
                int y = i / _terrain.Width;
                _terrain.SetMoisture(x, y, _terrain.Moisture[i] + gap * DryRate);
            }
        }

        private bool IsCellOverWater(int cx, int cy)
        {
            var (x, y) = CellCentreTile(cx, cy);
            return _terrain.IsWater(x, y);
        }

        private (int x, int y) CellCentreTile(int cx, int cy) => (
            System.Math.Min(_terrain.Width - 1, cx * WeatherCell + WeatherCell / 2),
            System.Math.Min(_terrain.Height - 1, cy * WeatherCell + WeatherCell / 2));

        /// <summary>Snapshot pour la sauvegarde (v2).</summary>
        public WeatherState Serialize() => new WeatherState
        {
            Cloud = (float[])Cloud.Clone(),
            WindAngle = WindAngle,
            AdvectionX = _advectionX,
            AdvectionY = _advectionY,
            RngState = _rng.GetState(),
        };

        /// <summary>Restauration depuis une sauvegarde (v2).</summary>
        public void Restore(WeatherState data)
        {
            if (data.Cloud.Length != Cloud.Length)
                throw new InvalidOperationException("Corrupted save: weather grid size mismatch");
            Array.Copy(data.Cloud, Cloud, Cloud.Length);
            WindAngle = data.WindAngle;
            _advectionX = data.AdvectionX;
            _advectionY = data.AdvectionY;
            _rng.SetState(data.RngState);
        }
    }

    /// <summary>État sérialisé de la météo (sauvegarde v2).</summary>
    public struct WeatherState
    {
        public float[] Cloud;
        public double WindAngle;
        public double AdvectionX;
        public double AdvectionY;
        public uint RngState;
    }
}
